app/controllers/api/avatars.php

```diff
@@ -49,11 +53,92 @@
 };
 
+$getUserGitHub = function (string $userId, Document $project, Database $dbForProject, Database $dbForConsole) {
+    try {
+        $user = Authorization::skip(fn () => $dbForConsole->getDocument('users', $userId));
+
+        $sessions = $user->getAttribute('sessions', []);
+        $session = $sessions[0] ?? new Document();
+
+        $provider = $session->getAttribute('provider');
+        $accessToken = $session->getAttribute('providerAccessToken');
+        $accessTokenExpiry = $session->getAttribute('providerAccessTokenExpiry');
+        $refreshToken = $session->getAttribute('providerRefreshToken');
+
+        $appId = $project->getAttribute('authProviders', [])[$provider . 'Appid'] ?? '';
+        $appSecret = $project->getAttribute('authProviders', [])[$provider . 'Secret'] ?? '{}';
+
+        $className = 'Appwrite\\Auth\\OAuth2\\' . \ucfirst($provider);
+
+        if (!\class_exists($className)) {
+            throw new Exception(Exception::PROJECT_PROVIDER_UNSUPPORTED);
+        }
+
+        $oauth2 = new $className($appId, $appSecret, '', [], []);
+
+        $isExpired = new \DateTime($accessTokenExpiry) < new \DateTime('now');
+        if ($isExpired) {
+            try {
+                $oauth2->refreshTokens($refreshToken);
+
+                $accessToken = $oauth2->getAccessToken('');
+                $refreshToken = $oauth2->getRefreshToken('');
+
+                $verificationId = $oauth2->getUserID($accessToken);
+
+                if (empty($verificationId)) {
+                    throw new \Exception("Locked tokens."); // Race codition, handeled in catch
+                }
+
+                $session
+                    ->setAttribute('providerAccessToken', $accessToken)
+                    ->setAttribute('providerRefreshToken', $refreshToken)
+                    ->setAttribute('providerAccessTokenExpiry', DateTime::addSeconds(new \DateTime(), (int)$oauth2->getAccessTokenExpiry('')));
+
+                Authorization::skip(fn () => $dbForProject->updateDocument('sessions', $session->getId(), $session));
+
+                $dbForProject->deleteCachedDocument('users', $user->getId());
+            } catch (Throwable $err) {
+                $index = 0;
+                do {
+                    $previousAccessToken = $session->getAttribute('providerAccessToken');
+
+                    $user = Authorization::skip(fn () => $dbForConsole->getDocument('users', $userId));
+                    $sessions = $user->getAttribute('sessions', []);
+                    $session = $sessions[0] ?? new Document();
+                    $accessToken = $session->getAttribute('providerAccessToken');
+
+                    if ($accessToken !== $previousAccessToken) {
+                        break;
+                    }
+
+                    $index++;
+                    \usleep(500000);
+                } while ($index < 10);
+            }
+        }
+
+        $oauth2 = new $className($appId, $appSecret, '', [], []);
```

Comment:
```suggestion
```
it's being reinitialised ?

Reply:
Yes, on purpose, to lose all the user data from previous check that is intended for locking concurrent requests. By re-initializing we make sure that user is gone, and we can safely fetch new user data with our safe access token